sb/sb.py

```diff
@@ -0,0 +1,95 @@
+import math
+from bit_vector import BitVector
+
+
+def gen_sb(width: int,
+           num_tracks: int,
+           sides: int,
+           feedthrough_outputs: str,
+           registered_outputs: str,
+           pe_output_count: int,
+           is_bidi: int,
+           sb_fs: str):
+    CONFIG_DATA_WIDTH = 32
+    CONFIG_ADDR_WIDTH = 32
+    feedthrough_count = feedthrough_outputs.count("1")
+    registered_count = registered_outputs.count("1")
+    outputs_driven_per_side = num_tracks - feedthrough_count
+    mux_height = (sides - 1) + pe_output_count
+    config_bit_count_per_output = math.ceil(math.log(mux_height, 2))
+    config_bit_count_per_side = config_bit_count_per_output*outputs_driven_per_side       # nopep8
+    config_bit_count_total = config_bit_count_per_side*sides
+    if registered_count > 0:
+        config_bit_count_total += registered_count*sides
+    num_config_regs = math.ceil(config_bit_count_total / CONFIG_DATA_WIDTH)
+    reset_val = BitVector(0, CONFIG_DATA_WIDTH)
+
+    class SB:
+        def __init__(self):
+            self.config = [BitVector(0, CONFIG_DATA_WIDTH)
+                           for _ in range(num_config_regs)]
+            self.reset()
+
+        def reset(self):
+            self.configure(BitVector(0, CONFIG_ADDR_WIDTH), reset_val)
+            self.read_data = None
+            self.out = [[None for j in range(num_tracks)] for i in range(sides)]
+
+        def configure(self, addr: BitVector, data: BitVector):
+            addr_high_bits = addr[24:32]
+            config_reg_select = addr_high_bits.as_uint()
+            if config_reg_select in range(num_config_regs):
+                self.config[config_reg_select] = data
+
+        # Function to slice the global config bit space. Returns bits in the
+        # range [lo, hi).
+        def get_config_bits(self, lo: int, hi: int):
```

Comment:
I believe this is the function causing a decreased coverage. Let's remove it -- we should either put it in a common place since the cb fn model uses it...or even better when we move away from this bit level configuration, we can avoid this altogether. For now let's remove the function.

Reply:
Have removed this function for now.